### PR TITLE
refactor(ChangeStream): use maybePromise for close, improve tests

### PR DESCRIPTION
**Pull request by emadum**
## Description

[NODE-2598](https://jira.mongodb.org/browse/NODE-2598)

**What changed?**

 - Refactor `close` to use `maybePromise` helper
 - Ensure explicitly closed change streams don't process further changes
 - Fix broken `mid-close` tests by delaying the final write
 - Add `waitForStarted` helper to tests where appropriate to reduce flakiness

**Are there any files to ignore?**

### CODE REVIEW

File: lib/change_stream.js [hunk 1]
```diff
@@ -158,7 +160,7 @@ class ChangeStream extends EventEmitter {
     if (this.cursor) {
       return this.cursor.isClosed();
     }
-    return true;
+    return this.closed;
```
Comment (mbroadst) [inline lib/change_stream.js:163]: I think we should _only_ return `this.close`, instead of deferring to the underlying cursor.

File: lib/change_stream.js [hunk 2]
```diff
@@ -168,31 +170,21 @@ class ChangeStream extends EventEmitter {
...
   close(callback) {
-    if (!this.cursor) {
-      if (callback) return callback();
-      return this.promiseLibrary.resolve();
-    }
-
-    // Tidy up the existing cursor
-    const cursor = this.cursor;
+    return maybePromise(this.parent, callback, cb => {
+      this.closed = true;
```
Comment (mbroadst) [inline lib/change_stream.js:174]:
```
if (this.closed) return cb();
```

File: lib/change_stream.js [hunk 3]
```diff
@@ -168,31 +170,21 @@ class ChangeStream extends EventEmitter {
...
 
-    if (callback) {
-      return cursor.close(err => {
-        ['data', 'close', 'end', 'error'].forEach(event => cursor.removeAllListeners(event));
-        delete this.cursor;
+      if (!this.cursor) {
```
Comment (mbroadst) [inline lib/change_stream.js:176]: I think you can remove this line after the above change is made

File: lib/change_stream.js [hunk 4]
```diff
@@ -168,31 +170,21 @@ class ChangeStream extends EventEmitter {
...
-    }
+      // Tidy up the existing cursor
+      const cursor = this.cursor;
 
-    const PromiseCtor = this.promiseLibrary || Promise;
-    return new PromiseCtor((resolve, reject) => {
-      cursor.close(err => {
+      return cursor.close(err => {
         ['data', 'close', 'end', 'error'].forEach(event => cursor.removeAllListeners(event));
         delete this.cursor;
```
Comment (mbroadst) [inline lib/change_stream.js:185]: nit: `this.cursor = undefined;` (I know you didn't introduce this change)

File: test/functional/change_stream.test.js [hunk 5]
```diff
@@ -2038,17 +2036,19 @@ describe('Change Streams', function() {
...
-            changeStream.next(err => {
-              let _err = null;
-              try {
-                expect(err.message).to.equal('ChangeStream is closed');
-              } catch (e) {
-                _err = e;
-              } finally {
-                done(_err);
-              }
+            changeStream.close(err => {
```
Comment (mbroadst) [inline test/functional/change_stream.test.js:2039]: this doesn't seem to jive with the name of the test, it seems like it more wants something like a race between `close` and `next`, e.g.
```
changeStream.next(() => {
  changeStream.close(() => {
  });

  changeStream.next(err => {
    expect(err).to.exist; // ...
    done();
  });
});
```

File: lib/change_stream.js [hunk 6]
```diff
@@ -556,8 +525,7 @@ function processNewChange(args) {
     );
 
     if (eventEmitter) return changeStream.emit('error', noResumeTokenError);
-    if (typeof callback === 'function') return callback(noResumeTokenError, null);
-    return changeStream.promiseLibrary.reject(noResumeTokenError);
+    return callback(noResumeTokenError, null);
```
Comment (mbroadst) [inline lib/change_stream.js:528]: nit since we're here anyway: `return callback(noResumeTokenError);` (no `null`). This also applies to a number of these callbacks here.

File: lib/change_stream.js [hunk 7]
```diff
@@ -506,46 +495,26 @@ function processNewChange(args) {
...
-          resolve();
-        });
-      })
-        .then(
-          () => (changeStream.cursor = createChangeStreamCursor(changeStream, cursor.resumeOptions))
-        )
-        .then(() => changeStream.next());
+        changeStream.cursor = createChangeStreamCursor(changeStream, cursor.resumeOptions);
+        if (eventEmitter) return;
+        changeStream.next(callback);
```
Comment (mbroadst) [inline lib/change_stream.js:511]: I think we still have a bug here if the thing that causes the recreate is `hasNext` (I think we discussed this the other day), I think a safer replacement would be to call `hasNext` here since that should cache the results.

File: lib/change_stream.js [hunk 8]
```diff
@@ -483,11 +475,8 @@ function processNewChange(args) {
     if (eventEmitter) {
       return;
     }
-
-    const error = new MongoError('ChangeStream is closed');
-    return typeof callback === 'function'
-      ? callback(error, null)
-      : changeStream.promiseLibrary.reject(error);
+    callback(new MongoError('ChangeStream is closed.'));
```
Comment (mbroadst) [inline lib/change_stream.js:478]: nit: no period at the end of the message (acknowledging that you didn't introduce this)

File: lib/change_stream.js (hunk 1 above)
Comment (emadum) [reply lib/change_stream.js:163]: I tried to do this, but the test `Should invalidate change stream on database drop using imperative callback form` starts failing. The underlying cursor closes due to the `invalidate` event, but `ChangeStream` doesn't have any way of knowing that except calling `this.cursor.isClosed()`. I suspect there's a way around that, but the current approach seems to work well enough.

File: test/functional/change_stream.test.js (hunk 5 above)
Comment (emadum) [reply test/functional/change_stream.test.js:2039]: Yeah, I suspected that was the case, but then I don't really understand the reasoning behind this test. Without the change I made, the test flakes out because everything works; the change event gets processed before the close. However the test requires that the close happen before the change event - why? Wouldn't it be better it if the change event happens before the close?  I suspect I'm missing something here.

File: lib/change_stream.js (hunk 7 above)
Comment (emadum) [reply lib/change_stream.js:506]: I was planning on deferring any logical changes to `ChangeStream` unrelated to the new `closed` property to an immediate follow-up PR to this, for [NODE-2548](https://jira.mongodb.org/browse/NODE-2548). Since `processNewChange` was using `next` before, I'd like to keep that for this PR; changing that to `hasNext` isn't in this one's scope. LMK if you disagree!

File: test/functional/change_stream.test.js [hunk 9]
```diff
@@ -2017,11 +2017,7 @@ describe('Change Streams', function() {
           return Promise.resolve()
             .then(() => changeStream.next())
             .then(() => changeStream.next())
-            .then(() => {
-              const nextP = changeStream.next();
-
-              return changeStream.close().then(() => nextP);
-            });
+            .then(() => Promise.all([changeStream.close(), changeStream.next()]));
```
Comment (mbroadst) [inline test/functional/change_stream.test.js:2020]: This is a behavioral change, it used to be that `next` was called (executing soon), and then the `close` call was made, after which next was awaited. Now you're essentially racing both `close` and `next`, and `close` comes first. 

Was there something you're working around here?

File: lib/change_stream.js (hunk 1 above)
Comment (mbroadst) [reply lib/change_stream.js]: Hmm, okay yeah that makes sense. If the cursor receives a `getMore` response with `cursorId: 0` then it will close the cursor but we won't find out about that until we call this. We'd only be able to track the `close` event, but that would force the cursor into streaming mode.

File: test/functional/change_stream.test.js (hunk 9 above)
Comment (emadum) [reply test/functional/change_stream.test.js:2020]: I thought I wrote a comment about that but I don't see it anymore, odd. 

> it used to be that next was called (executing soon), and then the close call was made, after which next was awaited

The test fails under those conditions. The final next returns the change when it is awaited, and the test expects an error. I can only get it to return an error if I call `close` first. I'm still not exactly sure what the desired behavior is that this test is trying to confirm happens.

Comment (mbroadst) [reply test/functional/change_stream.test.js:2020]: Here's the pull request where it was introduced, for more context: https://github.com/mongodb/node-mongodb-native/pull/1902

I think the idea is that we want to test the behavior when a new change comes in while the user is actively closing the change stream. It might just be an edge case that was possible before some fix that you introduced in this PR? It also seems like a test that could only really be handled through mocking or clever use of fail points, do you think we should remove the test?

Comment (emadum) [reply test/functional/change_stream.test.js:2020]: Yeah this is all making more sense now. I think the difference in behavior is derived from [this change](https://github.com/mongodb/node-mongodb-native/pull/2294/files#diff-30861ceb96e5170099331504d187951eR489), at which point they started flaking out until I skipped them in #2322. 

If I'm understanding correctly, this test was checking to make sure that `processNewChange` would throw an error if the underlying cursor closed. However, that shouldn't really be the case - the underlying cursor closes during a resumable error, but the stream shouldn't error.

So yeah, if that's correct I'm all for removing the test.

Comment (mbroadst) [reply test/functional/change_stream.test.js:2020]: I _think_ I agree, but now I'm wondering more about the `isClosed` dilemma also addressed in this PR. `processNewChange` has:
```
  if (cursor == null) {
    changeStream.closed = true;
    if (eventEmitter) {
      return;
    }

    callback(new MongoError('ChangeStream is closed'));
    const error = new MongoError('ChangeStream is closed');
    return;
  }
```

But what happens if `changeStream.closed === true` (i.e. the change stream was explicitly closed)? Shouldn't that trigger this path as well? Is that the behavior this test is testing?

Comment (emadum) [reply test/functional/change_stream.test.js:2020]: `next` already checks if `changeStream.closed === true` and returns an error before calling `processNewChange`. I confirmed by adding this to `processNewChange` and the test suite ran with no errors.
```
  if (changeStream.closed) {
    throw new Error('Should never get here');
  }
```

Comment (emadum) [reply test/functional/change_stream.test.js:2020]: That said, I'm thinking I should flip the or condition in `isClosed`.

If it was `return this.closed || (this.cursor && this.cursor.isClosed());` instead of the other way around, `closed === true` would take precedence over the cursor state.

Comment (mbroadst) [reply test/functional/change_stream.test.js:2020]: `next` checks this before calling `next` on the cursor, which in turn calls `processNewChange`. What if the change stream is not closed when you call `next`, but is closed sometime before the result of `cursor.next` returns? You might be able to test this with a mock server (or a sinon mock if we had a more complex fixture in place):
```
server.setResponseHandler(res => {
   const doc = res.doc;
   if (doc.ismaster) {
     res.reply({ .. ismaster .. });
   } else if (doc.aggregate) {
     res.reply({ ok: 1, .. other aggregate stuff .. });
   } else if (doc.getMore) {
     setTimeout(() => res.reply({ ok: 1, .. other getmore stuff .. }, 200); // always respond with latency to `getMore` calls
   }
});
```

now what does this do:
```
const changeStream = coll.watch();
changeStream.next(err => { // first call will send `aggregate`
  expect(err).to.not.exist;

  changeStream.next(err => {  // second call will send `getMore`
    expect(err).to.exist; // I think we _do_ expect to get an error here, change stream was closed explicitly. 
    done();
  });

  changeStream.close(); // immediately close the change stream
});
```

It may be that the underlying cursor logic already protects us from this, but I think that this is the behavior the test is meaning to cover. I just want to make sure we're not losing that coverage here.

Comment (emadum) [reply test/functional/change_stream.test.js:2020]: That makes sense, I'll check into it.

File: test/functional/change_stream.test.js [hunk 10]
```diff
@@ -2609,34 +2619,24 @@ describe('Change Streams', function() {
   describe('tryNext', function() {
```
Comment (emadum) [inline test/functional/change_stream.test.js:2619]: I oddly started getting an error in the `tryNext` tests:

`MongoError: cannot open $changeStream for non-existent database: test`

I thought it would be more robust to use a specifically named database and drop it after each test, to ensure weird behavior doesn't carry over between tests. To that purpose I created the `withTempDb` helper that wraps `withClient` to create and drop the specified database. This refactor made above error go away.

File: test/functional/change_stream.test.js (hunk 9 above)
Comment (emadum) [reply test/functional/change_stream.test.js]: I'm thinking it would make sense to create a separate ticket to fix those two `mid-close` tests. They were already skipped before this PR, so keeping them skipped until a followup PR doesn't hurt IMO. Thoughts, @mbroadst?

File: test/functional/change_stream.test.js [hunk 11]
```diff
@@ -37,7 +38,11 @@ function triggerResumableError(changeStream, onCursorClosed) {
  * @param {function} callback
  */
 function waitForStarted(changeStream, callback) {
+  const timeout = setTimeout(() => {
+    throw new Error('Change stream never started');
+  }, 2000);
```
Comment (emadum) [inline test/functional/change_stream.test.js:43]: This _shouldn't_ do anything. I'm just concerned some tests are timing out after 60s because for some reason this never triggers. If that's the case, this will at least fail a lot sooner, and more importantly we'll know for sure whether this is the root cause of test flakiness.

File: test/functional/change_stream.test.js [hunk 12]
```diff
@@ -54,9 +59,12 @@ function tryNext(changeStream, callback) {
   let complete = false;
   function done(err, result) {
     if (complete) return;
-
     // if the arity is 1 then this a callback for `more`
     if (arguments.length === 1) {
+      if (err instanceof Error) {
```
Comment (mbroadst) [inline test/functional/change_stream.test.js:64]: in what case is an error returned here? When I look at where `more` is emitted in `change_stream.js`, the errors are always handled. This would seem to imply the `getMore` callback was called with `err, err` or something.

File: test/functional/change_stream.test.js [hunk 13]
```diff
@@ -2609,34 +2638,24 @@ describe('Change Streams', function() {
...
-        client.connect(err => {
-          expect(err).to.not.exist;
-
-          const changeStream = client
-            .db()
-            .collection('test')
-            .watch();
-
+      test: function() {
+        return withTempDb('testTryNext', this.configuration.newClient(), db => done => {
```
Comment (mbroadst) [inline test/functional/change_stream.test.js:2642]: thoughts about the helpers:
- can `withClient` be modified to not need the `this.configuration.newClient` parameter if `this.configuration` exists?
```
withClient((client, done) => {
...
});
```
- I'm a little wary of functions that return functions (ad infinitum) in cases like this because it makes the test a bit hard to reason about. To be clear, I think what happens behind the scenes is totally fair game, but what we expose to the user (us) should be super intuitive, something I jump to use. What about composability:
```
withClient(withTempDb('testTryNext', (client, db, done) => {
  ... test ...
}));
```

File: test/functional/shared.js [hunk 14]
```diff
@@ -75,6 +75,17 @@ function makeCleanupFn(client) {
   };
 }
 
+function withTempDb(dbName, client, operation, errorHandler) {
+  return withClient(
+    client,
+    client => done => {
+      const db = client.db(dbName);
+      operation.call(this, db)(() => db.dropDatabase(done));
```
Comment (mbroadst) [inline test/functional/shared.js:83]: I imagine the reason you ran into this in the first place is because the database wasn't created with a wc: majority. We might want to ensure the same here

File: test/functional/change_stream.test.js (hunk 13 above)
Comment (emadum) [reply test/functional/change_stream.test.js:2642]: I agree it's worth thinking these helpers through a bit more. I had to make some modifications to `withMonitoredClient` in #2354 which could also be more elegant/usable.

File: test/functional/change_stream.test.js (hunk 12 above)
Comment (emadum) [reply test/functional/change_stream.test.js:64]: Oops I think I should remove this now, I was just trying to track down an issue but it's no longer necessary.
